File: 1414-find-the-minimum-number-of-fibonacci-numbers-whose-sum-is-k/1414-find-the-minimum-number-of-fibonacci-numbers-whose-sum-is-k.cpp
class Solution {
public:
    int findMinFibonacciNumbers(int k) {
        // Generate Fibonacci numbers up to k
        vector<int> fibs = {1, 1};
        while (fibs.back() <= k) {
            fibs.push_back(fibs.back() + fibs[fibs.size() - 2]);
        }
        
        int count = 0;
        int i = fibs.size() - 1;
        
        // Use the largest possible Fibonacci number and subtract it from k
        while (k > 0) {
            if (fibs[i] <= k) {
                k -= fibs[i];
                count++;
            }
            i--;
        }
        
        return count;
    }
};

File: 1414-find-the-minimum-number-of-fibonacci-numbers-whose-sum-is-k/1414-find-the-minimum-number-of-fibonacci-numbers-whose-sum-is-k.rb
# @param {Integer} k
# @return {Integer}
def find_min_fibonacci_numbers(k)
  # Generate Fibonacci numbers up to k
  fibs = [1, 1]
  while fibs[-1] <= k
    fibs << fibs[-1] + fibs[-2]
  end
  
  count = 0
  i = fibs.length - 1
  
  # Use the largest possible Fibonacci number and subtract it from k
  while k > 0
    if fibs[i] <= k
      k -= fibs[i]
      count += 1
    end
    i -= 1
  end
  
  count
end

